lib/galaxy/datatypes/molecules.py

```diff
@@ -674,6 +795,689 @@ def set_peek(self, dataset):
+        """
+        Find Atom IDs for metadata.
+        """
+        self.meta_error = False
+        if ase_io:
+            # enhanced metadata
+            try:
+                ase_data = ase_io.read(dataset.file_name, format="castep-cell")
+            except Exception as e:
+                log.warning('%s, set_meta Exception during ASE read: %s', self, unicodify(e))
+                self.meta_error = True
+                return
+
+            try:
+                atom_data = [
+                    str(sym) + str(pos) for sym, pos in zip(ase_data.get_chemical_symbols(), ase_data.get_positions())
+                ]
+                chemical_formula = ase_data.get_chemical_formula()
+                pbc = ase_data.get_pbc()
+                lattice_parameters = ase_data.get_cell().cellpar()
+            except Exception as e:
+                log.warning('%s, set_meta Exception during ASE metadata collection: %s', self, unicodify(e))
+                self.meta_error = True
+                return
+
+            # CELL file can only have one molecule
+            dataset.metadata.number_of_molecules = 1
+            dataset.metadata.atom_data = atom_data
+            dataset.metadata.number_of_atoms = len(dataset.metadata.atom_data)
+            dataset.metadata.chemical_formula = chemical_formula
+            try:
+                dataset.metadata.is_periodic = 1 if pbc else 0
+            except ValueError:  # pbc is an array
+                dataset.metadata.is_periodic = 1 if pbc.any() else 0
+            dataset.metadata.lattice_parameters = list(lattice_parameters)
+
+        else:
+            # simple metadata
+            with open(dataset.file_name) as f:
+                cell = f.read()
+            try:
+                # atom position data follows this pattern:
+                # %BLOCK POSITIONS(_FRAC|_ABS)
+                # [position data, one line per atom]
+                # %ENDBLOCK POSITIONS(_FRAC|_ABS)
+                block = (
+                    re.search(
+                        r"\n%BLOCK POSITIONS([\s\S]*?)\n%ENDBLOCK POSITIONS",
+                        cell,
+                        flags=re.IGNORECASE,
+                    )
+                    .group(1)
+                    .split("\n")[1:]
+                )
+                atom_data = []
+                for atom in block:
+                    atom = atom.split()
+                    symbol = atom[0].lower().capitalize()
+                    position = [float(i) for i in atom[1:4]]
+                    atom_data.append(symbol + str(position))
+
+                # CELL file can only have one molecule
+                dataset.metadata.number_of_molecules = 1
+                dataset.metadata.atom_data = atom_data
+                dataset.metadata.number_of_atoms = len(dataset.metadata.atom_data)
+            except Exception as e:
+                log.warning('%s, set_meta Exception during simple metadata collection: %s', self, unicodify(e))
```

Comment:
I think you're right that an all-or-nothing approach to metadata is more sensible. We have tools that will work fine without the metadata so no need to raise an exception if it's unset.

Is there any way to print the warning/error somewhere so that it can be seen in the logs or ideally preserved with the job?